File: pluginbase-core/messages/src/main/java/pluginbase/messages/DefaultMessageProvider.java
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package pluginbase.messages;

import ninja.leaping.configurate.ConfigurationNode;
import ninja.leaping.configurate.loader.ConfigurationLoader;
import org.jetbrains.annotations.Nullable;
import pluginbase.logging.PluginLogger;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.Arrays;
import java.util.IllegalFormatException;
import java.util.Iterator;
import java.util.Locale;

class DefaultMessageProvider implements MessageProvider {

    @NotNull
    private final Locale locale;

    @NotNull
    private final ConfigurationNode messages;

    @NotNull
    private final LocalizablePlugin plugin;

    public DefaultMessageProvider(@NotNull final LocalizablePlugin localizablePlugin,
                                  @NotNull final ConfigurationLoader loader,
                                  @NotNull final Locale locale) {
        this.locale = locale;
        this.plugin = localizablePlugin;
        messages = load(localizablePlugin, loader);
        prune(localizablePlugin, messages);
        save(loader, messages);
    }

    private void save(@NotNull final ConfigurationLoader loader, @NotNull final ConfigurationNode language) {
        try {
            loader.save(language);
        } catch (IOException e) {
            getLog().warning("Problem saving current language file");
            e.printStackTrace();
        }
    }

    private void prune(@NotNull final LocalizablePlugin localizable, @NotNull final ConfigurationNode language){
        // Prune file
        final Iterator<? extends ConfigurationNode> it = language.getChildrenList().iterator();
        while (it.hasNext()) {
            final Object[] key = it.next().getPath();
            if (!Messages.containsMessageKey(localizable, key)) {
                getLog().finer("Removing unused language: %s", key);
                it.remove();
            }
        }
    }

    @NotNull
    private ConfigurationNode load(@NotNull final LocalizablePlugin localizable, @NotNull final ConfigurationLoader loader) {
        ConfigurationNode language;
        try {
            language = loader.load();
        } catch (IOException e) {
            e.printStackTrace();
            language = loader.createEmptyNode();
        }

        // TODO check if languages is empty?

        for (final Object[] key : Messages.getMessageKeys(localizable)) {
            final Message message = Messages.getMessage(localizable, key);
            ConfigurationNode node = language.getNode(key);
            if (message != null && node.isVirtual()) {
                if (node.isVirtual()) {
                    node.setValue(message.getDefault());
                    getLog().finest("Created new message in language file: %s", message);
                } else if(Message.countArgs(node.getString())
                        != message.getArgCount()) {
                    node.setValue(message.getDefault());
                    getLog().warning("The message for '%s' in the current language file does not have the correct amount of arguments (%s).  The default will be used.", key, message.getArgCount());
                }
            }
        }
        return language;
    }

    private String _getMessage(@NotNull final Object[] key) {
        ConfigurationNode node = this.messages.getNode(key);
        if (node.isVirtual()) {
            String keyString = Arrays.toString(key);
            getLog().warning("There is not language entry for %s.  Was it registered?", keyString);
            return keyString;
        }
        return node.getString();
    }

    @Override
    @NotNull
    public String getLocalizedMessage(@NotNull final Message key, @NotNull final Object... args) {
        if (key.getKey() == null) {
            return formatMessage(null, key.getDefault(), args);
        }
        return getLocalizedMessage(key.getKey(), args);
    }

    @NotNull
    @Override
    public String getLocalizedMessage(@NotNull final Object[] key, @NotNull final Object... args) {
        final String message = _getMessage(key);
        return formatMessage(key, message, args);
    }

    private String formatMessage(@Nullable Object[] key, @NotNull String message, @NotNull final Object... args) {
        try {
            return MessageUtil.formatMessage(locale, message, args);
        } catch (IllegalFormatException e) {
            getLog().warning("Language string format is incorrect: %s: %s", key, message);
            for (final StackTraceElement ste : Thread.currentThread().getStackTrace()) {
                getLog().warning(ste.toString());
            }
            return message;
        }
    }

    @NotNull
    @Override
    public LocalizablePlugin getPlugin() {
        return plugin;
    }

    @NotNull
    public PluginLogger getLog() {
        return getPlugin().getLog();
    }
}

File: pluginbase-core/messages/src/main/kotlin/pluginbase/messages/DefaultMessageProvider.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
package pluginbase.messages

import ninja.leaping.configurate.ConfigurationNode
import ninja.leaping.configurate.loader.ConfigurationLoader
import pluginbase.logging.PluginLogger

import java.io.IOException
import java.util.Arrays
import java.util.IllegalFormatException
import java.util.Locale

internal class DefaultMessageProvider(override val plugin: LocalizablePlugin,
                                      loader: ConfigurationLoader<*>,
                                      private val locale: Locale) : MessageProvider {

    private val messages: ConfigurationNode

    init {
        messages = load(plugin, loader)
        prune(plugin, messages)
        save(loader, messages)
    }

    private fun save(loader: ConfigurationLoader<*>, language: ConfigurationNode) {
        try {
            loader.save(language)
        } catch (e: IOException) {
            log.warning("Problem saving current language file")
            e.printStackTrace()
        }

    }

    private fun prune(localizable: LocalizablePlugin, language: ConfigurationNode) {
        // Prune file
        val it = language.childrenList.iterator()
        while (it.hasNext()) {
            val key = it.next().path
            if (!Messages.containsMessageKey(localizable, key)) {
                log.finer("Removing unused language: %s", *key)
                it.remove()
            }
        }
    }

    private fun load(localizable: LocalizablePlugin, loader: ConfigurationLoader<*>): ConfigurationNode {
        var language: ConfigurationNode
        try {
            language = loader.load()
        } catch (e: IOException) {
            e.printStackTrace()
            language = loader.createEmptyNode()
        }

        // TODO check if languages is empty?

        for (key in Messages.getMessageKeys(localizable)) {
            val message = Messages.getMessage(localizable, key)
            val node = language.getNode(*key)
            if (message != null && node.isVirtual) {
                if (node.isVirtual) {
                    node.value = message.default
                    log.finest("Created new message in language file: %s", message)
                } else if (Message.countArgs(node.string) != message.argCount) {
                    node.value = message.default
                    log.warning("The message for '%s' in the current language file does not have the correct amount of arguments (%s).  The default will be used.", key, message.argCount)
                }
            }
        }
        return language
    }

    private fun _getMessage(key: Array<Any>): String {
        val node = this.messages.getNode(*key)
        if (node.isVirtual) {
            val keyString = Arrays.toString(key)
            log.warning("There is not language entry for %s.  Was it registered?", keyString)
            return keyString
        }
        return node.string
    }

    override fun getLocalizedMessage(key: Message, vararg args: Any?): String {
        if (key.key == null) {
            return formatMessage(null, key.default, *args)
        }
        return getLocalizedMessage(key.key, *args)
    }

    override fun getLocalizedMessage(key: Array<Any>, vararg args: Any?): String {
        val message = _getMessage(key)
        return formatMessage(key, message, *args)
    }

    private fun formatMessage(key: Array<Any>?, message: String, vararg args: Any?): String {
        try {
            return formatMessage(locale, message, *args)
        } catch (e: IllegalFormatException) {
            log.warning("Language string format is incorrect: %s: %s", key, message)
            // Help identify where the incorrect language elements are
            for (ste in Thread.currentThread().stackTrace) {
                log.warning(ste.toString())
            }
            return message
        }

    }

    val log: PluginLogger
        get() = plugin.log
}
